Adding stateHash function

nanspro opened the pull request:
Implementing #4

johannbarbie commented:
@nanspro good start. thanks for the effort 👍 

A few comments from looking at the hash function. I think due to the nature of solEVM as script executor (EVM in EVM), the following attributes of the VM will not be used:
- evm.value,  `// as solEVM has no global state, there is also no Ether`
- evm.caller.nonce, `//  as solEVM has no global state, there is also no nonce`
- evm.caller.destroyed, `// accounts can not be destroyed`
- evm.caller.code, `// transactions can only come from accounts`
- evm.staticExec, `// all executions are non-static, but can only affect limited scope`
- evm.context.difficulty `// no difficulty in the simulated environment solEVM is running in`

The following elements I don't understand. Maybe you can explain them to me:
- evm.stack.dataPtr,
- evm.caller.stge.size,

Same question, what is the difference between those:
- evm.target.code and evm.code

nanspro commented:
- I went through .slb files and i saw that `dataPtr` only changes in some function and instruction codes of memory and stack never uses those kind of operations hence there is no need of dataPtr

- No instruction code changes `EVMStorage.storage.size`  

@johannbarbie do we need `caller` and `target` fields?
`evmInput.caller = DEFAULT_CALLER;
 evmInput.target = DEFAULT_CONTRACT_ADDRESS; `
They are never changed(No instruction code can change account balance) so i don;'t think they are required for hashing.

nanspro commented:
@johannbarbie  i have made some changes, could you review the hash function again?

nanspro commented:
- I went through .slb files and i saw that `dataPtr` only changes in some function and instruction codes of memory and stack never uses those kind of operations hence there is no need of dataPtr most of the time. But in some cases like stack.push() when stack size is equal to stack cap then it will call `expandCapacity(self, 32);` which will change the dataPtr so should i hash the dataPtr always to include these cases.

- No instruction code changes `EVMStorage.storage.size`  

@johannbarbie do we need `caller` and `target` fields?
`evmInput.caller = DEFAULT_CALLER;
 evmInput.target = DEFAULT_CONTRACT_ADDRESS; `
They are never changed(No instruction code can change account balance) so i don't think they are required for hashing.

johannbarbie commented on test/onChain.test.js:
```diff
@@ -37,8 +37,12 @@ contract('Runtime', function () {
+        console.log('On', onstateHash);
         assert.deepEqual(toStr(onChainState.stack), toStr(afterState.stack), 'Stack');
-        assert.equal(onChainState.mem, afterState.mem, 'Memory');
+        // assert.equal(onstateHash, offstateHash, 'State Hash');
```
the idea is to replace all the specific comparisons with only the comparison of the hashes.

you could even change the runtime to only return a hash from `executeAndStop()` and `initAndExecute()`

nanspro replied:
we can't return only hash because we want `beforeState` for `initAndExecute` to obtain `onState`. 
`const beforeState = await rt.executeAndStop(code, callData, [0, step, BLOCK_GAS_LIMIT, BLOCK_GAS_LIMIT])`
`const onState = await rt.initAndExecute(
          code, callData,
          [beforeState.pc, 1, BLOCK_GAS_LIMIT, beforeState.gasRemaining],
          beforeState.stack, beforeState.mem, beforeState.accounts, beforeState.accountsCode,
          beforeState.logHash,
        );`

johannbarbie commented:
> do we need caller and target fields?
i guess we could leave the out.
@pinkiebell what do you think?

nanspro commented:
In onChaintest.js `const offstateHash = rt.stateHash(afterState, 2, 1);` rt uses function which are in runtimeAdapter.js and it doesn't have `stateHash()` so how to use `stateHash()` for tests, we can write one more `stateHash()` in runtimeAdapter.js and but we'll have to generate parameters for this function?

johannbarbie commented:
> so how to use stateHash() for tests

why not return replace this with your function? https://github.com/leapdao/solEVM-enforcer/blob/master/contracts/EthereumRuntime.sol#L139-L144

pinkiebell commented:
> > do we need caller and target fields?
> 
> i guess we could leave them out.
> @pinkiebell what do you think?

Yes, we can leave everything with accounts out for now.

nanspro commented:
@johannbarbie 
I have removed the `flattenResult()` and we are using execute to return `hashState` which unlike `Result` contains all the parameters for hash function and also `hashValue`.
For `initAndExecute()` we need `beforeState` to pass the parameters so we need state along hash to compare.

nanspro commented:
@johannbarbie  i have added some parameters to `Result` and changed its name to `hashState` because it's attribute are sufficient for hashing and one attribute is the `hashValue`.

johannbarbie commented:
@nanspro thanks for the rework. checking. could you fix the merge conflicts in the mean time?

nanspro commented:
@johannbarbie conflicts has been resolved!

pinkiebell commented on contracts/EthereumRuntime.sol:
```diff
@@ -278,6 +306,58 @@ contract EthereumRuntime is EVMConstants, IEthereumRuntime {
+            //evm.context,
+
+            evm.mem.size,
+            evm.mem.cap,
```
You can drop `.cap` here, as it is implementation specific.

pinkiebell commented on contracts/EthereumRuntime.sol:
```diff
@@ -278,6 +306,58 @@ contract EthereumRuntime is EVMConstants, IEthereumRuntime {
+            // There is no OP_CODE which can interact with setCapacity and no OP_CODE can change the dataPtr.
+
+            evm.stack.size,
+            evm.stack.cap,
```
Same as above^

pinkiebell commented on contracts/Verifier.sol:
```diff
@@ -8,6 +8,8 @@ import "./IEthereumRuntime.sol";
 import "./Merkelizer.slb";
 import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
 
+import { IEthereumRuntime } from "./IEthereumRuntime.sol";
```
Already import above :)

pinkiebell commented on contracts/EthereumRuntime.sol:
```diff
@@ -85,6 +85,24 @@ contract EthereumRuntime is EVMConstants, IEthereumRuntime {
         bytes32 logHash;
     }
 
+    // struct HashParameters {
```
You can remove these comments if that structure isn't used somewhere

pinkiebell commented on test/onChain.test.js:
```diff
@@ -31,22 +32,23 @@ contract('Runtime', function () {
-        assert.deepEqual(onChainState.accounts, afterState.accounts, 'Accounts');
-        assert.equal(onChainState.accountsCode, afterState.accountsCode, 'Accounts code');
-        assert.equal(onChainState.logHash, afterState.logHash, 'Log hash');
+        // assert.deepEqual(toStr(onChainState.stack), toStr(afterState.stack), 'Stack');
```
Do they need to be commented out?

nanspro replied:
If we are checking `hashValue` then we are also equating stack, memory so `hashValue` should be enough to compare states.

pinkiebell commented:
> Also one more thing about parameters, when i went through .slb files and i saw that `dataPtr` (for stack and memory both) only changes in some function and instruction codes of memory and stack never uses those kind of operations hence there is no need of `dataPtr` most of the time. But in some cases like `stack.push()` when stack size is equal to stack cap then it will call `expandCapacity(self, 32)` which will change the `dataPtr` so should i hash the `dataPtr` always to include these cases.

Nope, you don't have to hash dataPtr, but you actually need to hash the size/length of stack,memory.
Note: Not the `.cap` :)

pinkiebell commented:
@nanspro Great! You can submit on Gitcoin 👍

peara commented on contracts/EthereumRuntime.sol:
```diff
@@ -112,7 +112,7 @@ contract EthereumRuntime is EVMConstants, IEthereumRuntime {
     // Init EVM with given stack and memory and execute from the given opcode
     // solhint-disable-next-line function-max-lines
-    function execute(EVMPreimage memory img) public pure returns (EVMPreimage memory) {
+    function execute(EVMPreimage memory img) public pure returns (EVMResult memory) {
```
Why `EVMResult` is needed here?
I think most of its attributes is the same as `EVMPreimage`?
I think `stateHash(EVMPreimage)` function should be enough?

nanspro replied:
@peara
`EVMResult` 's attributes are parameters of EVM which are passed down to stateHash so i thought instead of using the earlier `result` (whose parameters were random) we may use a result with better info.
Yes `stateHash(EVMPreimage)` will be enough but do we want the same struct to behave both as input and output of functions?

peara replied:
I think using one struct could has benefits such as I can pass it to the next `execute`. but as things are hashed, we can't do that.
So most of the info is redundant actually..
I don't know for sure, but maybe just return hash could be better

pinkiebell replied:
From my point of view, you can leave it as is for now. Because some things will change, that isn't going to be a problem anyway 🙂. We will probably use custom functions to invoke the EVM in different parts of the code, so leaving it as is should be fine for the tests/fixtures.

pinkiebell commented:
manually merged; resolved conflicts
